packages/js/onboarding/src/components/WooOnboardingTaskReferral/index.ts

```diff
@@ -0,0 +1,71 @@
+type StoredItem< T > = {
+	data: T;
+	expiry: string;
+};
+
+/**
+ * This constructor function exists for type safety purposes, so when you create a pair of set/get functions they have the
+ *  same type for the value they store/retrieve.
+ */
+export function createStorageUtils< T >(
+	key: string,
+	durationInSeconds: number,
+	type: 'session' | 'local' = 'local'
+) {
+	const storage = type === 'session' ? sessionStorage : localStorage;
+	return {
+		setWithExpiry: ( value: T ): void => {
+			const now = new Date();
+			const expiry = new Date( now.getTime() + durationInSeconds * 1000 );
+
+			const item: StoredItem< T > = {
+				data: value,
+				expiry: expiry.toISOString(),
+			};
+
+			storage.setItem( key, JSON.stringify( item ) );
+		},
+		getWithExpiry: (): T | null => {
+			const itemStr = storage.getItem( key );
+			if ( ! itemStr ) {
+				return null;
+			}
+
+			const item: StoredItem< T > = JSON.parse( itemStr );
+			const now = new Date();
+			if ( now > new Date( item.expiry ) ) {
+				storage.removeItem( key );
+				return null;
+			}
+			return item.data;
+		},
+		remove: (): void => {
+			storage.removeItem( key );
+		},
+	};
+}
+
+export type TaskReferralRecord = {
+	referrer: string;
+	returnUrl: string;
+};
+
+/**
+ * Returns a pair of getter and setter for the task referral storage.
+ *
+ * @param {Object} options                  - The options object.
+ * @param {string} options.taskId           - Name of the task.
+ * @param {number} options.referralLifetime - Lifetime of the referral. Default of 60 seconds is a reasonable amount of time for referral handoff lifetime in most cases, since it really only needs to account for the page load duration
```

Comment:
Does it mean a user only has 60s to complete the task? 🤔

Reply:
Got it, thanks! 

24 hours is probably fine, but does throw up an edge-case where a user might close their computer for the day and return back to it a few days later. 7 days might be better but it's not a blocker for me.